### PR TITLE
Created a testable network layer 

### PR DESCRIPTION
**Pull request by sandesh-nk**

### CODE REVIEW

File: iOS-TemplateTests/APITests/GitUserAPITests.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        }
+        
+        let configuration = URLSessionConfiguration.default
+        configuration.protocolClasses = [URLProtocolMock.self]
+        let api = GitUsersAPI(configuration)
+        api.searchUsers(query: "sandesh", page: 1) { (result) in
+            switch result {
+            case .success(let model):
+                expectation.fulfill()
+                XCTAssertGreaterThanOrEqual(model.items.count, 6)
```
Comment (rameez-wed) [inline iOS-TemplateTests/APITests/GitUserAPITests.swift]: can't we just use XCTAssertEqual?

File: iOS-TemplateTests/GitubEndpointTests.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        XCTAssertEqual(endPoint.scheme, "https")
+        XCTAssertEqual(endPoint.baseURL, "api.github.com")
+        XCTAssertEqual(endPoint.path, "/search/users")
+    }
+    
+    func testRepoSearchURLGeneration() {
+        let endPoint = GithubEndpoint.repos(username: "sandesh")
+        XCTAssertEqual(endPoint.scheme, "https")
+        XCTAssertEqual(endPoint.baseURL, "api.github.com")
+        XCTAssertEqual(endPoint.path, "/users/sandesh/repos")
```
Comment (rameez-wed) [inline iOS-TemplateTests/GitubEndpointTests.swift]: avoid multiple assertions in the test. Create a separate test for the each assertion you need to make

**Review by rameez-wed — CHANGES_REQUESTED** [review]

File: iOS-Template/Networking/API's/ITunesSongsAPI.swift [hunk 3]
```diff
@@ -0,0 +1,17 @@
+//
+//  ITunesSongsAPI.swift
+//  iOS-Template
+//
+//  Created by Apple on 20/10/21.
+//
+
+import Foundation
+
+struct ITunesSongsAPI {
```
Comment (rameez-wed) [inline iOS-Template/Networking/API's/ITunesSongsAPI.swift:10]: this is your network engine right not an API. Rename to something meaningful.

File: iOS-Template/Networking/Networking.swift [hunk 4]
```diff
@@ -8,84 +8,50 @@
...
-      self.currentNetworkRequest?.cancel()
+        
+        var urlRequest = URLRequest(url: url)
+        urlRequest.httpMethod = endpoint.method
+        let session = URLSession(configuration: configuration)
+        
+        print(url.absoluteString)
+        
+        let dataTask = session.dataTask(with: urlRequest) { (data, response, error) in
+            guard error == nil else {
```
Comment (rameez-wed) [inline iOS-Template/Networking/Networking.swift]: if let to unwrap the error would make sense instead of checking for nil

File: iOS-TemplateTests/APITests/ITunesSongsAPITests.swift [hunk 5]
```diff
@@ -0,0 +1,19 @@
+//
+//  ITunesSongsAPITests.swift
+//  iOS-TemplateTests
+//
+//  Created by Apple on 20/10/21.
+//
+
+import XCTest
+
+class ITunesSongsAPITests: XCTestCase {
```
Comment (rameez-wed) [inline iOS-TemplateTests/APITests/ITunesSongsAPITests.swift]: delete if it has no tests

File: iOS-TemplateTests/ITunesEndPointTests.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+//  ITunesEndPointTests.swift
+//  iOS-TemplateTests
+//
+//  Created by Apple on 20/10/21.
+//
+
+import XCTest
+import iOS_Template
+
+class ITunesEndPointTests: XCTestCase {
```
Comment (rameez-wed) [inline iOS-TemplateTests/ITunesEndPointTests.swift:11]: Please add relevant tests or delete if there are none

File: iOS-Template/Networking/Endpoints/ITunesEndpoint.swift [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        switch self {
+        default:
+            return "itunes.apple.com"
+        }
+    }
+    
+    var path: String {
+        switch self {
+        case .getSong:
+            return "search"
```
Comment (rameez-wed) [inline iOS-Template/Networking/Endpoints/ITunesEndpoint.swift]: You will need /search or it will failed when you generate url from components

**Review by rameez-wed — APPROVED** [review]
looks good
